### PR TITLE
Permanent Volumes for Kubernetes deployment

### PR DESCRIPTION
**Pull request by MarkvanMents**
Add information on permanent volumes for Kubernetes deployments as discussed in issue  #4853
Also proofread for style

### CODE REVIEW

File: content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md [hunk 1]
```diff
@@ -8,46 +8,51 @@ tags: ["Kubernetes", "cloud", "deployment"]
...
+If the application makes use of persistable FileDocument or FileImage entities, a persistent volume (PV) storage service needs to be attached as well. See [Mendix System Requirements](/refguide/system-requirements/) for supported external storage services. In this how-to you use a node-bound storage volume as an example. For more information, see [Architecture Overview](#architecture), below.
 
 This how-to uses [Minikube](https://kubernetes.io/docs/getting-started-guides/minikube/), which is a way to run Kubernetes locally. Many of the operations you perform on Minikube are the same as those on a hosted environment and it provides a low-level entry to Kubernetes. For more information, see [Installing Kubernetes with Minikube](https://kubernetes.io/docs/setup/learning-environment/minikube/) on the Kubernetes documentation site.
 
 For more details on Kubernetes, see the [Kubernetes Documentation](https://kubernetes.io/docs/home/) site.
 
+{{% todo %}}Are these configuration files really available? Better just to copy and paste?{{% /todo %}}
+
 All the configuration files used in this how-to are also available on GitHub.
```
Comment (MarkvanMents) [inline content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md]: Don't know where these files are? Can't see anything obvious in mendix/docs or mendix/docker-mendix-buildpack.
I think we should remove this and just suggest copying and pasting from the code blocks in the document?

File: content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md [hunk 2]
```diff
@@ -8,46 +8,51 @@ tags: ["Kubernetes", "cloud", "deployment"]
...
+* Install kubectl using the instructions provided in [Install and Set Up kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl)
     * The kubectl CLI is the default tool to access and manage your Kubernetes cluster
-    * Install kubectl based on the instructions provided in [Install and Set Up kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl)
-* Install Minikube
-    * With Minikube, a local cluster can be created that is convenient for exploring Kubernetes (if you have an account for one of the cloud providers and you choose to use that, this step can be skipped)
-    * Install Minikube based on the instructions provided in [Install Minikube](https://kubernetes.io/docs/tasks/tools/install-minikube/)
+* Install Minikube using the instructions provided in [Install Minikube](https://kubernetes.io/docs/tasks/tools/install-minikube/)
+    * Minikube allows you to create a local cluster that is convenient for exploring Kubernetes (if you have an account for one of the cloud providers and you choose to use that, this step can be skipped)
+* Build your image in Minikube, using the steps in the [Docker](/developerportal/deploy/docker-deploy/) page.
```
Comment (MarkvanMents) [inline content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md]: Moved this to a prerequisite as there were no instructions on how to do it, it just said it had to be done first.
Is there anything else we need to say here?

File: content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md [hunk 3]
```diff
@@ -61,28 +66,25 @@ The deployment of your Mendix app needs the following Kubernetes components:
...
 ### 4.1 Deploying the PostgreSQL Database
 
-Once Minikube is running you'll need to configure your local environment to use the Docker daemon using the following command:
+Once Minikube is running you must configure your local environment to use the Docker daemon using the following command:
 
 ```bash {linenos=false}
 minikube docker-env
 ```
-
-You'll need to build your image in Minikube if you haven't done so yet. See [Docker](/developerportal/deploy/docker-deploy/) for the steps to do this.
```
Comment (MarkvanMents) [inline content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md:80]: Moved to prerequisite.

File: content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md [hunk 4]
```diff
@@ -155,41 +157,174 @@ The database is now created. To verify the installation, check out the logs:
...
 ```
 
-### 4.2 Deploying the Application{#deploy}
+### 4.2 Adding a Persistent Volume
+
+The Docker Buildpack stores files in /opt/mendix/build/data/files. If you do not have any persistent storage, then these files will disappear if the pod is destroyed. If you mount a Persistent Volume (PV) into this path, any uploaded files will be stored in that PV.
+
+To do this, create the `buildpack-openshift.yaml` with the content below:
+
+```yaml
```
Comment (MarkvanMents) [inline content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md]: Is this the right thing to do.
I think the yaml does more than just add persistent volumes. However, it doesn't match exactly with the instructions on database configuration and deployment/stateful set in the other yaml.
Is there any way to pull out what is relevant to persistent volumes and indicate what the reader should do with them?

File: content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md [hunk 5]
```diff
@@ -8,46 +8,51 @@ tags: ["Kubernetes", "cloud", "deployment"]
...
 {{% /alert %}}
 
 A Mendix application needs, as a minimum, a database to run. In this example you provision a PostgreSQL database within the Kubernetes cluster. In production scenarios, the database is usually provided as a service by the cloud provider, like AWS RDS or Azure SQL. For supported databases see [Mendix System Requirements](/refguide/system-requirements/). 
 
-If the application makes use of FileDocument or FileImage entities, a storage service needs to be attached as well. See [Mendix System Requirements](/refguide/system-requirements/) for supported external storage services. In this how-to you use a node-bound storage volume as an example. For more information, see [Architecture Overview](#architecture), below.
+If the application makes use of persistable FileDocument or FileImage entities, a persistent volume (PV) storage service needs to be attached as well. See [Mendix System Requirements](/refguide/system-requirements/) for supported external storage services. In this how-to you use a node-bound storage volume as an example. For more information, see [Architecture Overview](#architecture), below.
```
Comment (MarkvanMents) [inline content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md:21]: Is it true that we are still using node-bound storage with the new section we have added about adding persistent storage?

**Review by MarkvanMents — COMMENTED** [review]
I've added a few comments on things about which I was unsure.
@zlogic and @elliotclements-mendix, can you look at this and see if it solves the issues or just raises more.
Because this is written as a how-to, the idea is that the customer can have a working deployment at the end of it. Therefore, the instructions for persistent storage should be executable within the context of the how to.

Let me know your thoughts.
Thanks.
Mark

File: content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md (hunk 1 above)
Comment (zlogic) [reply content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md]: I think the files are available in https://github.com/mendix/kubernetes-howto, but they look very outdated. Indeed, it would be better to remove any references to that repository.

File: content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md (hunk 2 above)
Comment (zlogic) [reply content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md]: The document doesn't mention it, but most likely the user would need to use [minikube build](https://minikube.sigs.k8s.io/docs/commands/image/#minikube-image-build) instead of `docker build`.

File: content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md [hunk 6]
```diff
@@ -8,46 +8,51 @@ tags: ["Kubernetes", "cloud", "deployment"]
 
 ## 1 Introduction
 
-This how-to describes what is needed to deploy your Mendix app to [Kubernetes](https://kubernetes.io/). Kubernetes is the standard Docker orchestration platform supported by Mendix. If possible, we suggest you use [Mendix for Private Cloud](/developerportal/deploy/private-cloud/) to deploy Mendix apps to Kubernetes as this provides you with integration with the Developer Portal and takes away some of the heavy lifting. For details on supported version of Kubernetes see [Mendix System Requirements](/refguide/system-requirements/).
+This how-to takes you through the process of deploying your Mendix app to [Kubernetes](https://kubernetes.io/).
+
+Kubernetes is the standard Docker orchestration platform supported by Mendix. For details on supported version of Kubernetes see [Mendix System Requirements](/refguide/system-requirements/). If possible, we suggest you use [Mendix for Private Cloud](/developerportal/deploy/private-cloud/) to deploy Mendix apps to Kubernetes as this provides you with integration with the Developer Portal and takes away some of the heavy lifting.
```
Comment (zlogic) [inline content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md]: Could you rephrase this as _Kubernetes is the standard container orchestration platform..._

Kubernetes and Docker parted ways (https://kubernetes.io/blog/2022/05/03/dockershim-historical-context/), they share the same basic principles and can use the same container images, but otherwise they don't depend on each other.

File: content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md (hunk 5 above)
Comment (zlogic) [reply content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md:21]: The example provided below needs an extra component, a [volume provisioner](https://github.com/kubernetes-sigs/sig-storage-local-static-provisioner/blob/master/docs/provisioner.md) to use node-bound storage. Otherwise the app will not start until a local volume provisioner is installed and configured. I've provided a comment how to use node-bound storage directly, bypassing a volume provisioner below.

File: content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md [hunk 7]
```diff
@@ -204,15 +339,15 @@ data:
...
 kubectl get ep postgres-service
 ```
 
 See [Run a Mendix Docker Image](/developerportal/deploy/run-mendix-docker-image/) for expected value formats.
 
-To create the secrets in Kubernetes we execute the following command:
+Create the secrets in Kubernetes by executing the following command:
 
 ```bash {linenos=false}
 kubectl create -f mendix-app-secrets.yaml
```
Comment (zlogic) [inline content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md:224]: I think there's an easier way to fix host (node)-bound storage, just change
```
          volumeMounts:
            - mountPath: "/build/data/files"
              name: mendix-data
```
to
```
          volumeMounts:
            - mountPath: "/opt/mendix/build/data/files"
              subPath: files
              name: mendix-data
```

in this example `StatefulSet`.

File: content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md (hunk 4 above)
Comment (zlogic) [reply content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md]: I think the StatefulSet example below is out of date and is using incorrect paths - so this section is not required.

**Review by zlogic — CHANGES_REQUESTED** [review]

File: content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md (hunk 6 above)
Comment (MarkvanMents) [reply content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md]: I'm a bit confused about this as the documentation refers to the "Docker Buildpack".
Is there a way we can make this a bit clearer?
We can pick this up in a separate change if it is too much to do it now.

File: content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md (hunk 2 above)
Comment (MarkvanMents) [reply content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md]:
```suggestion
* Build your image in Minikube, using the steps in the [Docker](/developerportal/deploy/docker-deploy/) page and replacing Docker commands such as `docker build`  with Minikube equivalents such as `minikube image build`.
```

File: content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md (hunk 4 above)
Comment (MarkvanMents) [reply content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md]: Removed

File: content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md (hunk 7 above)
Comment (MarkvanMents) [reply content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md:224]: Updated, see below

File: content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md (hunk 5 above)
Comment (MarkvanMents) [reply content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md:21]: I assume the changes I made to the `mendix-app.yaml` cover everything we need to do here?

File: content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md [hunk 8]
```diff
@@ -283,13 +290,13 @@ Create a Docker image of your Mendix app using the instructions in the [Mendix D
 Once you have created the Docker image, push it to the Docker hub using the following command:
 
 ```bash {linenos=false}
-docker push <hub-user>/<repo-name>:<tag>
+docker push <hub-user>/<repo-name>:<tag>
```
Comment (MarkvanMents) [inline content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md]: Should this be `minikube push` or is `docker push` OK?

**Review by MarkvanMents — COMMENTED** [review]
@zlogic 
Hi Dmitrii,
I've updated the files from your comments. Can you let me know if this is now OK or whether there is anything further that needs doing.
Thanks for your quick review.
Mark

File: content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md (hunk 6 above)
Comment (zlogic) [reply content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md]: If I understand correcly, originally Kubernetes users would build app images using Docker Buildpack and then deploy them into Kubernetes using the YAMLs from this document. Mx4PC is intended to provide a higher level interface (a web UI for Connected mode or a MendixApp manifest for Standalone mode). But when this document was originally written, Mx4PC didn't exist yet and so the only way to deploy apps into Kubernetes was by using Docker Buildpack.

File: content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md (hunk 2 above)
Comment (zlogic) [reply content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md]: I made a mistake, can you change `minikube build` to `minikube image build`?

File: content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md (hunk 8 above)
Comment (zlogic) [reply content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md]: I've double-checked, and the command should be:
* `minikube image push <hub-user>/<repo-name>:<tag>` instead of `docker push ...`
* `minikube image build` instead of `docker build` or `minikube build`

**Comment by zlogic** [comment]
@MarkvanMents Looks good to me, just one comment from my side - instead of `docker build` `docker push`, the command should be `minikube image build` and `minikube image push`.

File: content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md (hunk 2 above)
Comment (MarkvanMents) [reply content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md]: Edited. Does that look better?

File: content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md (hunk 8 above)
Comment (MarkvanMents) [reply content/en/docs/developerportal/deploy/docker-deploy/run-mendix-on-kubernetes.md]: So this?
```suggestion
minikube image push <hub-user>/<repo-name>:<tag>
```

**Review by MarkvanMents — COMMENTED** [review]
@zlogic 
Thanks for the review.
Let me know if the two proposals are now correct and I'll commit them and merge (unless you have any other concerns.

**Comment by MarkvanMents** [comment]
As discussed, will merge now - if there is anything else we can pick it up later, but this is an improvement on the current document!
Thanks for your help Dmitrii (@zlogic )
